src/Common/WebSocketExtensions.cs

```diff
@@ -1,42 +1,60 @@
 {
     internal static class WebSocketExtensions
     {
-        public static Task SendAsync(this WebSocket webSocket, ReadOnlySequence<byte> buffer, WebSocketMessageType webSocketMessageType, CancellationToken cancellationToken = default)
+        public static ValueTask SendAsync(this WebSocket webSocket, ReadOnlySequence<byte> buffer, WebSocketMessageType webSocketMessageType, CancellationToken cancellationToken = default)
         {
-            // TODO: Consider chunking writes here if we get a multi segment buffer
 #if NETCOREAPP2_1
             if (buffer.IsSingleSegment)
             {
-                return webSocket.SendAsync(buffer.First, webSocketMessageType, endOfMessage: true, cancellationToken).AsTask();
+                return webSocket.SendAsync(buffer.First, webSocketMessageType, endOfMessage: true, cancellationToken);
             }
             else
             {
-                return webSocket.SendAsync(buffer.ToArray(), webSocketMessageType, endOfMessage: true, cancellationToken);
+                return SendMultiSegmentAsync(webSocket, buffer, webSocketMessageType, cancellationToken);
             }
 #else
             if (buffer.IsSingleSegment)
             {
                 var isArray = MemoryMarshal.TryGetArray(buffer.First, out var segment);
                 Debug.Assert(isArray);
-                return webSocket.SendAsync(segment, webSocketMessageType, endOfMessage: true, cancellationToken);
+                return new ValueTask(webSocket.SendAsync(segment, webSocketMessageType, endOfMessage: true, cancellationToken));
             }
             else
             {
-                return webSocket.SendAsync(new ArraySegment<byte>(buffer.ToArray()), webSocketMessageType, true, cancellationToken);
+                return SendMultiSegmentAsync(webSocket, buffer, webSocketMessageType, cancellationToken);
             }
+#endif
+        }
+
+        private static async ValueTask SendMultiSegmentAsync(WebSocket webSocket, ReadOnlySequence<byte> buffer, WebSocketMessageType webSocketMessageType, CancellationToken cancellationToken = default)
+        {
+            var position = buffer.Start;
+            while (buffer.TryGet(ref position, out var segment))
+            {
+#if NETCOREAPP2_1
+                await webSocket.SendAsync(segment, webSocketMessageType, endOfMessage: false, cancellationToken);
+#else
+                var isArray = MemoryMarshal.TryGetArray(segment, out var arraySegment);
+                Debug.Assert(isArray);
+                await webSocket.SendAsync(arraySegment, webSocketMessageType, endOfMessage: false, cancellationToken);
+#endif
+            }
+
+            // Empty end of message frame
+#if NETCOREAPP2_1
+            await webSocket.SendAsync(Memory<byte>.Empty, webSocketMessageType, endOfMessage: true, cancellationToken);
+#else
+            await webSocket.SendAsync(new ArraySegment<byte>(Array.Empty<byte>()), webSocketMessageType, endOfMessage: true, cancellationToken);
```

Comment:
Cache this?

Reply:
It's a struct so I don't think there is any benefit